File: lib/rgfa/field/alignment_gfa1.rb
module RGFA::Field::AlignmentGFA1

  def decode(string)
    string.to_cigar(valid: false, version: :"1.0")
  end

  def unsafe_decode(string)
    string.to_cigar(valid: true, version: :"1.0")
  end

  def validate_encoded(string)
    if string !~ /^(\*|([0-9]+[MIDNSHPX=])+)$/
      raise RGFA::FormatError,
        "#{string.inspect} is not a valid GFA1 alignment\n"+
        "(it is not * and is not a CIGAR string (([0-9]+[MIDNSHPX=])+)"
    end
  end

  def validate_decoded(object)
    case object
    when RGFA::Alignment::CIGAR
      object.validate
    when RGFA::Alignment::Placeholder
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: String, RGFA::Alignment::CIGAR,"+
        "RGFA::Alignment::Placeholder)"
    end
  end

  def unsafe_encode(object)
    object.to_s
  end

  def encode(object)
    case object
    when String
      validate_encoded(object)
      return object
    when RGFA::Alignment::CIGAR
      object.validate
      return object.to_s
    when RGFA::Alignment::Placeholder
      return object.to_s
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: String, "+
        "RGFA::Alignment::CIGAR, RGFA::Alignment::Placeholder)"
    end
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :unsafe_encode
  module_function :encode

end

File: gfapy/field/alignment_gfa1.py
import gfapy
import re

def decode(string):
  return gfapy.Alignment(string, valid = False, version = "gfa1")

def unsafe_decode(string):
  return gfapy.Alignment(string, valid = True, version = "gfa1")

def validate_encoded(string):
  if not re.match(r"^(\*|([0-9]+[MIDNSHPX=])+)$", string):
    raise gfapy.FormatError(
      "{} is not a valid GFA1 alignment\n".format(repr(string)) +
      "(it is not * and is not a CIGAR string (([0-9]+[MIDNSHPX=])+)")

def validate_decoded(obj):
  if isinstance(obj, gfapy.CIGAR):
    obj.validate()
  elif isinstance(obj, gfapy.Placeholder):
    pass
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name) +
      "(accepted classes: gfapy.CIGAR, gfapy.Placeholder)")

def unsafe_encode(obj):
  return str(obj)

def encode(obj):
  if isinstance(obj, str):
    validate_encoded(obj)
    return obj
  elif isinstance(obj, gfapy.CIGAR):
    obj.validate()
    return str(obj)
  elif isinstance(obj, gfapy.Placeholder):
    return str(obj)
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__) +
      "(accepted classes: str, gfapy.CIGAR, gfapy.Placeholder)")
